packages/tax/src/services/tax-module-service.ts

```diff
@@ -166,4 +167,145 @@ export default class TaxModuleService<
+      {
+        $or: [
+          {
+            country_code: calculationContext.address.country_code,
+            province_code: null,
+          },
+          {
+            country_code: calculationContext.address.country_code,
+            province_code: calculationContext.address.province_code,
+          },
+        ],
+      },
+      {},
+      sharedContext
+    )
+
+    const toReturn = await promiseAll(
+      items.map(async (item) => {
+        const isShipping = "shipping_option_id" in item
+        let ruleQuery = isShipping
+          ? [
+              {
+                reference: "shipping_option",
+                reference_id: item.shipping_option_id,
+              },
+            ]
+          : [
+              {
+                reference: "product",
+                reference_id: item.product_id,
+              },
+              {
+                reference: "product_type",
+                reference_id: item.product_type_id,
+              },
+            ]
+
+        const rates = await this.taxRateService_.list(
+          {
+            $and: [
+              { tax_region_id: { $in: regions.map((r) => r.id) } },
+              { $or: [{ is_default: true }, { rules: { $or: ruleQuery } }] },
+            ],
+          },
+          { relations: ["rules", "tax_region"] },
+          sharedContext
+        )
+
+        const prioritizedRates = this.prioritizeRates(rates, item)
+
+        const rate = prioritizedRates[0]
+        const toReturn = {
+          rate_id: rate.id,
+          rate: rate.rate,
+          code: rate.code,
+          name: rate.name,
+        }
+
+        if (isShipping) {
+          return [{ shipping_line_id: item.id, ...toReturn }]
+        }
+
+        return [{ line_item_id: item.id, ...toReturn }]
+      })
+    )
+
+    return toReturn.flat()
+  }
+
+  private prioritizeRates(
+    rates: TTaxRate[],
+    item: TaxTypes.TaxableItemDTO | TaxTypes.TaxableShippingDTO
+  ) {
+    const isShipping = "shipping_option_id" in item
+
+    const decoratedRates: (TTaxRate & {
+      priority_score: number
+    })[] = rates.map((rate) => {
+      let isProductMatch = false
+      let isProductTypeMatch = false
+      if (rate.rules.length !== 0) {
+        const matchingRules = rate.rules.filter((rule) => {
+          if (isShipping) {
+            return (
+              rule.reference === "shipping" &&
+              rule.reference_id === item.shipping_option_id
+            )
+          }
+          return (
+            (rule.reference === "product" &&
+              rule.reference_id === item.product_id) ||
+            (rule.reference === "product_type" &&
+              rule.reference_id === item.product_type_id)
+          )
+        })
+
+        if (matchingRules.length !== 0) {
+          if (matchingRules.some((rule) => rule.reference === "product")) {
+            isProductMatch = true
+          } else {
+            isProductTypeMatch = true
+          }
```

Comment:
could also be shipping here:

```suggestion
          } else if (matchingRules.some((rule) => rule.reference === "product_type")) {
            isProductTypeMatch = true
          }
```